libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +265,122 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
-	solAssert(!_contract.isLibrary(), "Tried to initialize state variables of library.");
-
-	using boost::adaptors::reverse;
-
-	ostringstream out;
-
-	FunctionDefinition const* constructor = _contract.constructor();
-	if (constructor && !constructor->isPayable())
-		out << callValueCheck();
+	map<ContractDefinition const*, string> constructorParams;
+	vector<pair<ContractDefinition const*, std::vector<ASTPointer<Expression>>>> baseConstructorArguments;
+	if (FunctionDefinition const* constructor = _contract.constructor())
+		for (auto const& modifier: constructor->modifiers())
+		{
+			FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+				modifier->name()->annotation().referencedDeclaration
+			)->constructor();
+
+			if (baseConstructor && modifier->arguments() && !modifier->arguments()->empty())
+				baseConstructorArguments.emplace_back(
+					dynamic_cast<ContractDefinition const*>(baseConstructor->scope()),
+					*modifier->arguments()
+				);
+		}
 
-	for (ContractDefinition const* contract: reverse(_contract.annotation().linearizedBaseContracts))
+	for (ASTPointer<InheritanceSpecifier> const& base: _contract.baseContracts())
 	{
-		out <<
-			"\n// Begin state variable initialization for contract \"" <<
-			contract->name() <<
-			"\" (" <<
-			contract->stateVariables().size() <<
-			" variables)\n";
-
-		IRGeneratorForStatements generator{m_context, m_utils};
-		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!variable->isConstant() && !variable->immutable())
-				generator.initializeStateVar(*variable);
-		out << generator.code();
-
-		out << "// End state variable initialization for contract \"" << contract->name() << "\".\n";
+		FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+			base->name().annotation().referencedDeclaration
+		)->constructor();
+
+		if (baseConstructor && base->arguments() && !base->arguments()->empty())
+			baseConstructorArguments.emplace_back(
+				dynamic_cast<ContractDefinition const*>(baseConstructor->scope()),
+				*base->arguments()
+			);
 	}
 
-	if (constructor)
-	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (pair<ContractDefinition const*, std::vector<ASTPointer<Expression>>> const& baseArgs: baseConstructorArguments)
+		if (baseArgs.first->constructor() && !baseArgs.second.empty())
+		{
+			string params;
+			for (size_t i = 0; i < baseArgs.second.size(); ++i)
+				params += (params.empty() ? "" : ", ") + generator.evaluateExpression(
+						*(baseArgs.second.at(i)),
+						*(baseArgs.first->constructor()->parameters()[i]->type())
+				).commaSeparatedList();
+			constructorParams[baseArgs.first] = params;
```

Comment:
This is what ends up in the parameter list and the name of the variable is unchanged except for the implicit constructor of the contract itself, right? If it is not too much work, I think it would be good if the Solidity name of the constructor parameter could somehow end up in the yul parameter names.

Reply:
I would need to add contract name and id as parameter name can be the same for two different constructors